File: sololink/flightcode/util/RcLock.cpp

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include "file_util.h"
#include "RcLock.h"

namespace RcLock
{

// must match rc_lock.py
static const char *ro_lockfile = "/mnt/rootfs.ro/etc/.rc_lock";
static const char *tmp_lockfile = "/tmp/.rc_lock";
static const char *tmp_unlockfile = "/tmp/.rc_unlock";

// locked - see if RC is locked
//
// Returns true if RC is locked, false if not locked.
bool locked(void)
{

    if (file_exists(tmp_unlockfile))
        return false;
    else if (file_exists(tmp_lockfile) || file_exists(ro_lockfile))
        return true;
    else
        return false;
}

// lock_version - lock RC due to version mismatch
void lock_version(void)
{
    file_touch(tmp_lockfile);
}

// unlock_version - unlock RC (versions match)
void unlock_version(void)
{
    unlink(tmp_lockfile);
}

// unlock_override - unlock RC
void unlock_override(void)
{
    file_touch(tmp_unlockfile);
}

} // namespace RcLock

File: sololink/flightcode/python/rc_lock.py
#!/usr/bin/env python

import errno
import os


# must match RcLock.cpp
ro_lockfile = "/mnt/rootfs.ro/etc/.rc_lock"
tmp_lockfile = "/tmp/.rc_lock"
tmp_unlockfile = "/tmp/.rc_unlock"


def locked():
    if os.path.isfile(tmp_unlockfile):
        return False
    elif os.path.isfile(tmp_lockfile) or \
         os.path.isfile(ro_lockfile):
        return True
    else:
        return False


def lock_version():
    with open(tmp_lockfile, 'a'): # "touch"
        os.utime(tmp_lockfile, None)


def unlock_version():
    try:
        os.unlink(tmp_lockfile)
    except OSError as e:
        if e.errno != errno.ENOENT:
            raise e


def unlock_override():
    with open(tmp_unlockfile, 'a'): # "touch"
        os.utime(tmp_unlockfile, None)
